staging/src/k8s.io/kubectl/pkg/kuberc/kuberc.go

```diff
@@ -0,0 +1,315 @@
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"k8s.io/kubectl/pkg/config/v1alpha1"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+const RecommendedKubeRCFileName = ".kuberc"
+
+var (
+	RecommendedConfigDir  = filepath.Join(homedir.HomeDir(), clientcmd.RecommendedHomeDir)
+	RecommendedKubeRCFile = filepath.Join(RecommendedConfigDir, RecommendedKubeRCFileName)
+)
+
+// PreferencesHandler is responsible for setting default flags
+// arguments based on user's kuberc configuration.
+type PreferencesHandler interface {
+	AddFlags(flags *pflag.FlagSet)
+	ApplyOverrides(rootCmd *cobra.Command, args []string, errOut io.Writer) error
+	ApplyAliases(rootCmd *cobra.Command, args []string, errOut io.Writer) ([]string, error)
+}
+
+// Preferences stores the kuberc file coming either from environment variable
+// or file from set in flag or the default kuberc path.
+type Preferences struct {
+	GetPreferencesFunc func(kuberc string) (*v1alpha1.Preferences, error)
+
+	aliases map[string]struct{}
+}
+
+// NewPreferences returns initialized Prefrences object.
+func NewPreferences() PreferencesHandler {
+	return &Preferences{
+		GetPreferencesFunc: DefaultGetPreferences,
+		aliases:            make(map[string]struct{}),
+	}
+}
+
+// AddFlags adds kuberc related flags into the command.
+func (p *Preferences) AddFlags(flags *pflag.FlagSet) {
+	if util.KubeRC.IsEnabled() {
+		flags.String("kuberc", "", "Path to the kuberc file to use for preferences.")
+	}
+}
+
+// ApplyOverrides finds the command and sets the defaulted flag values in kuberc.
+func (p *Preferences) ApplyOverrides(rootCmd *cobra.Command, args []string, errOut io.Writer) error {
+	if len(args) <= 1 {
+		return nil
+	}
+
+	if !util.KubeRC.IsEnabled() {
+		return nil
+	}
+
+	kubercPath := getExplicitKuberc(args)
+	kuberc, err := p.GetPreferencesFunc(kubercPath)
+	if err != nil {
+		return fmt.Errorf("kuberc error %w", err)
+	}
+
+	if kuberc == nil {
+		return nil
+	}
+
+	args = args[1:]
+	cmd, _, err := rootCmd.Find(args)
+	if err != nil {
+		fmt.Fprintf(errOut, "command not found %v\n", err)
+		return nil
+	}
+
+	for _, c := range kuberc.Spec.Overrides {
+		parsedCmds := strings.Split(c.Command, " ")
+		overrideCmd, _, err := rootCmd.Find(parsedCmds)
+		if err != nil {
+			// this may be referring to the alias command which is not initialized
+			// because the actual command is totally different.
+			return nil
+		}
+		if overrideCmd.Name() != cmd.Name() {
+			continue
+		}
+
+		if _, ok := p.aliases[cmd.Name()]; ok {
+			return fmt.Errorf("alias %s can not be overridden", cmd.Name())
+		}
+
+		// This function triggers merging the persistent flags in the parent commands.
+		_ = cmd.InheritedFlags()
+
+		for _, fl := range c.Flags {
+			// explicit flag usage has higher precedence than the kuberc default flag value.
+			// We should set the default flag values in kuberc, unless there is any in args.
+			if explicitFlagUse(fl.Name, args) {
```

Comment:
Is this call necessary?

For one, it doesn't take into account short flags (`-o` versus `--output`)

But I removed this line of code and it seemed to work as expected... I could set `-o yaml` on the command line and it took precedence over the override I had defined in my .kuberc file.

Reply:
This is the crucial comment and I'll drop my thoughts (or fixes) after completing my investigation.

Reply:
I removed explicitly flag check entirely, it seems redundant. But I'll chase down why I added it before. Thank you.